### ISSUE TITLE
`ParamSpec` erases `Optional` when `no-strict-optional` is enabled

### ISSUE THREAD

**Issue by koogoro** [python/mypy#15932]
**Bug Report**

The latest master will cause `ParamSpec` to replace parameters with type `Optional[T]` with type `T` when in a file with `no-strict-optional`. This is different from the behavior on 1.5.1. While this is equivalent within the same file, if a function affected by this issue is imported to a file without `no-strict-optional`, mypy will no longer let you pass `None` into that parameter. This behavior affects decorators such as `contextlib.contextmanager`.

**To Reproduce**

https://gist.github.com/mypy-play/dc17766e33ca6f4a1ec440257b0462ac

**Expected Behavior**

```
main.py:10: note: Revealed type is "def (arg: Union[builtins.int, None])"
Success: no issues found in 1 source file
```

**Actual Behavior**

```
main.py:10: note: Revealed type is "def (arg: builtins.int)"
Success: no issues found in 1 source file
```

**Your Environment**

I bisected the commits and found that this issue first occurs on https://github.com/python/mypy/commit/14418bc3d2c38b9ea776da6029e9d9dc6650b7ea.

**Comment by ilevkivskyi** [comment]
This is consistent with behavior for regular `TypeVar`s (since at least 1.0.0, but likely it is much older):
```python
from typing import Optional, TypeVar
from typing import Callable

def a(arg: Optional[int]) -> None: ...

T = TypeVar('T')
def b(f: Callable[[T], None]) -> Callable[[T], None]: ...

reveal_type(b(a))  # Revealed type is "def (builtins.int)"
```
 Unions were simplified during expansion for ages, and inconsistent behavior for `ParamSpec` was a bug.
